codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.aayushatharva.brotli4j.encoder.Encoder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToByteEncoder;
+
+/**
+ * Compress a {@link ByteBuf} with with the brotli format.
+ *
+ * See <a href="https://github.com/google/brotli">brotli</a>.
+ */
+public final class BrotliEncoder extends MessageToByteEncoder<ByteBuf> {
+
+    private final Encoder.Parameters parameters;
+
+    /**
+     * Create a new {@link BrotliEncoder} Instance with Quality set to 4.
+     */
+    public BrotliEncoder() {
+        parameters = new Encoder.Parameters().setQuality(4);
+    }
+
+    /**
+     * Create a new {@link BrotliEncoder} Instance
+     *
+     * @param parameters {@link Encoder.Parameters} Instance
+     */
+    public BrotliEncoder(Encoder.Parameters parameters) {
+        this.parameters = parameters;
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out) throws Exception {
+        // If ByteBuf is unreadable, then just ignore it.
+        if (!msg.isReadable()) {
+            return;
+        }
+
+        byte[] uncompressed = ByteBufUtil.getBytes(msg, msg.readerIndex(), msg.readableBytes(), false);
+        out.writeBytes(Encoder.compress(uncompressed, parameters));
```

Comment:
this this one works with `byte[]` ?

Reply:
Yes

Reply:
😢

Reply:
Can we at least use offset, Len here? This would allow us to at least not do a memory copy if we use a heap buffer as input in all cases

Reply:
I have already set the copy to `false`. Will that not work?

Reply:
```java
public static byte[] getBytes(ByteBuf buf, int start, int length, boolean copy) {
        int capacity = buf.capacity();
        if (isOutOfBounds(start, length, capacity)) {
            throw new IndexOutOfBoundsException("expected: " + "0 <= start(" + start + ") <= start + length(" + length
                    + ") <= " + "buf.capacity(" + capacity + ')');
        }

        if (buf.hasArray()) {
            int baseOffset = buf.arrayOffset() + start;
            byte[] bytes = buf.array();
            if (copy || baseOffset != 0 || length != bytes.length) {
                return Arrays.copyOfRange(bytes, baseOffset, baseOffset + length);
            } else {
                return bytes;
            }
        }

        byte[] bytes = PlatformDependent.allocateUninitializedArray(length);
        buf.getBytes(start, bytes);
        return bytes;
    }
```

Reply:
This will only not do the copy of the offset is 0 and the length is the same as the underlying array. Both of this is most likely not true when using the pooled allocator

Reply:
Can you give me a snippet code of what should I use?

Reply:
I basically asked for something like this:


```
out.writeBytes(Encoder.compress(msg.array(), msg.arrayOffset() + msg.readerIndex(), msg.readableBytes(), parameters));
```

But I just noticed that such an API is not supported atm. Maybe something to consider for the future to add:

https://github.com/hyperxpro/Brotli4j/blob/main/brotli4j/src/main/java/com/aayushatharva/brotli4j/encoder/Encoder.java#L47